src/View/Asset.php

```diff
@@ -0,0 +1,336 @@
+            $filepath = preg_replace(
+                '/^' . preg_quote(static::requestWebroot(), '/') . '/',
+                '',
+                urldecode($path)
+            );
+            $webrootPath = Configure::read('App.wwwRoot') . str_replace('/', DIRECTORY_SEPARATOR, $filepath);
+            if (file_exists($webrootPath)) {
+                return $path . '?' . filemtime($webrootPath);
+            }
+            $segments = explode('/', ltrim($filepath, '/'));
+            $plugin = Inflector::camelize($segments[0]);
+            if (Plugin::isLoaded($plugin)) {
+                unset($segments[0]);
+                $pluginPath = Plugin::path($plugin)
+                    . 'webroot'
+                    . DIRECTORY_SEPARATOR
+                    . implode(DIRECTORY_SEPARATOR, $segments);
+                if (file_exists($pluginPath)) {
+                    return $path . '?' . filemtime($pluginPath);
+                }
+            }
+        }
+
+        return $path;
+    }
+
+    /**
+     * Checks if a file exists when theme is used, if no file is found default location is returned
+     *
+     * ### Options:
+     *
+     * - `theme` Optional theme name
+     *
+     * @param string $file The file to create a webroot path to.
+     * @param array $options Options array.
+     * @return string Web accessible path to file.
+     */
+    public static function webroot(string $file, array $options = []): string
+    {
+        $options += ['theme' => null];
+        $requestWebroot = static::requestWebroot();
+
+        $asset = explode('?', $file);
+        $asset[1] = isset($asset[1]) ? '?' . $asset[1] : '';
+        $webPath = $requestWebroot . $asset[0];
+        $file = $asset[0];
+
+        $themeName = $options['theme'];
+        if ($themeName) {
+            $file = trim($file, '/');
+            $theme = static::inflectThemeName($themeName) . '/';
+
+            if (DIRECTORY_SEPARATOR === '\\') {
+                $file = str_replace('/', '\\', $file);
+            }
+
+            if (file_exists(Configure::read('App.wwwRoot') . $theme . $file)) {
+                $webPath = $requestWebroot . $theme . $asset[0];
+            } else {
+                $themePath = Plugin::path($themeName);
+                $path = $themePath . 'webroot/' . $file;
+                if (file_exists($path)) {
+                    $webPath = $requestWebroot . $theme . $asset[0];
+                }
+            }
+        }
+        if (strpos($webPath, '//') !== false) {
+            return str_replace('//', '/', $webPath . $asset[1]);
+        }
+
+        return $webPath . $asset[1];
+    }
+
+    /**
+     * Inflect the theme name to its underscored version.
+     *
+     * @param string $name Name of the theme which should be inflected.
+     * @return string Inflected name of the theme
+     */
+    protected static function inflectThemeName(string $name): string
+    {
+        return Inflector::underscore($name);
```

Comment:
Do we need this additional function call/indirection?

Reply:
I copied this from UrlHelper, I can remove it.